File: src/main/java/io/github/Earth1283/fixLag/OverloadChecker.java
package io.github.Earth1283.fixLag;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import java.util.List;
import java.util.stream.Collectors;

public class OverloadChecker {

    private final FixLag plugin;
    private final List<String> entitiesToDelete;
    private int overloadRadius;
    private int criticalEntityCount;

    public OverloadChecker(FixLag plugin, List<String> entitiesToDelete) {
        this.plugin = plugin;
        this.entitiesToDelete = entitiesToDelete;
        loadConfig();
    }

    public void loadConfig() {
        FileConfiguration config = plugin.getConfig();
        overloadRadius = config.getInt("overload-detection.radius", 50);
        criticalEntityCount = config.getInt("overload-detection.critical-entity-count", 30);
    }

    public void checkOverloads() {
        for (Player player : Bukkit.getOnlinePlayers()) {
            Location playerLocation = player.getLocation();
            World world = playerLocation.getWorld();
            if (world != null) { // check all worlds by default
                List<Entity> nearbyEntities = world.getNearbyEntities(playerLocation, overloadRadius, overloadRadius, overloadRadius).stream()
                        .filter(entity -> entitiesToDelete.contains(entity.getType().name().toUpperCase()))
                        .filter(entity -> !entity.hasPermission("fixlag.overload.exempt"))
                        .collect(Collectors.toList());

                if (nearbyEntities.size() > criticalEntityCount) {
                    String message = ChatColor.RED + "Warning! High number of targeted entities (" + nearbyEntities.size() + ") detected near you.";
                    for (Player staff : Bukkit.getOnlinePlayers()) {
                        if (staff.isOp() || staff.hasPermission("fixlag.overload.notify")) {
                            staff.sendMessage(message);
                        }
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/io/github/Earth1283/fixLag/managers/OverloadChecker.kt
package io.github.Earth1283.fixLag.managers

import io.github.Earth1283.fixLag.FixLag
import org.bukkit.Bukkit
import org.bukkit.scheduler.BukkitTask

class OverloadChecker(private val plugin: FixLag, private val entitiesToDelete: List<String>) {

    private var overloadRadius: Int = 50
    private var criticalEntityCount: Int = 30
    private var checkTask: BukkitTask? = null

    init {
        loadConfig()
    }

    fun loadConfig() {
        val config = plugin.config
        overloadRadius = config.getInt("overload-detection.radius", 50)
        criticalEntityCount = config.getInt("overload-detection.critical-entity-count", 30)
    }

    fun startChecking(delayTicks: Long, periodTicks: Long) {
        checkTask = Bukkit.getScheduler().runTaskTimer(plugin, Runnable {
            val onlinePlayers = ArrayList(Bukkit.getOnlinePlayers())

            for (player in onlinePlayers) {
                val playerUUID = player.uniqueId
                val playerLocation = player.location
                val world = playerLocation.world

                if (world != null) {
                    val nearbyEntities = world.getNearbyEntities(playerLocation, overloadRadius.toDouble(), overloadRadius.toDouble(), overloadRadius.toDouble())
                        .toList()

                    Bukkit.getScheduler().runTaskAsynchronously(plugin, Runnable {
                        val typeFilteredEntities = nearbyEntities.filter { entity ->
                            entitiesToDelete.contains(entity.type.name.uppercase())
                        }

                        Bukkit.getScheduler().runTask(plugin, Runnable {
                            val p = Bukkit.getPlayer(playerUUID)
                            if (p != null && p.isOnline) {
                                val finalCount = typeFilteredEntities.count { entity ->
                                    entity.isValid && !entity.hasPermission("fixlag.overload.exempt")
                                }

                                if (finalCount > criticalEntityCount) {
                                    val message = plugin.messageManager.getMessage("overload_warning", "<count>", finalCount.toString(), "<player>", p.name)
                                    for (staff in Bukkit.getOnlinePlayers()) {
                                        if (staff.isOnline && (staff.isOp || staff.hasPermission("fixlag.overload.notify"))) {
                                            staff.sendMessage(message)
                                        }
                                    }
                                }
                            }
                        })
                    })
                }
            }
        }, delayTicks, periodTicks)
    }

    fun stopChecking() {
        checkTask?.let {
            if (!it.isCancelled) {
                it.cancel()
            }
            checkTask = null
        }
    }
}
